examples/nn.dx

```diff
@@ -0,0 +1,227 @@
+' # Neural Networks
+
+include "plot.dx"
+
+' ## NN Prelude
+
+
+def relu (input : Float) : Float =
+  select (input > 0.0) input 0.0
+
+instance [Add a, Add b] Add (a & b)
+  add = \(a, b) (c, d). ( (a + c), (b + d))
+  sub = \(a, b) (c, d). ( (a - c), (b - d))
+  zero = (zero, zero)
+
+instance [VSpace a, VSpace b] VSpace (a & b)
+  scaleVec = \ s (a, b) . (scaleVec s a, scaleVec s b)
+
+def Layer (inp:Type) (out:Type) (param:Type) : Type=
+  {forward:(param -> inp -> out) &
+   init: Key -> param
+  }
+
+def forward (l:Layer i o p) (p : p) (x : i): o =
+  (getAt #forward l) p x
+
+def init (l:Layer i o p) (k:Key)  : p  =
+  (getAt #init l) k
+
+
+' ## Layers 
+
+' Dense layer
+
+def Dense (a:Type) (b:Type) : Type =
+  Layer (a=>Float) (b=>Float) ((a=>b=>Float) & (b=>Float))
+
+def dense (a:Type) (b:Type) : Dense a b =
+  {
+   forward = (\(weight, bias) x .
+               for j:b. (bias.j + sum for i :a. weight.i.j * x.i)),
```

Comment:
From the discussions in some other PRs I got the impression that one might want to use
```suggestion
               for j:b. (bias.j + sum view i:a. weight.i.j * x.i)),
```
here (and similarly in the CNN implementation) for more predictable performance.

Reply:
For what it's worth this particular optimization is actually the one that our compiler is consistently able to perform :) So let's keep the `for`!